src/app-layer-detect-proto.c

```diff
@@ -1419,38 +1418,24 @@ AppProto AppLayerProtoDetectGetProto(AppLayerProtoDetectThreadCtx *tctx, Flow *f
                     FLOW_RESET_PP_DONE(f, reverse_dir);
                 }
             }
-
-            /* HACK: if detected protocol is dcerpc/udp, we run PP as well
-             * to avoid misdetecting DNS as DCERPC. */
-            if (!(ipproto == IPPROTO_UDP && alproto == ALPROTO_DCERPC))
-                goto end;
-
-            pm_alproto = alproto;
-
-            /* fall through */
+            SCReturnUInt(alproto);
         }
     }
 
     if (!FLOW_IS_PP_DONE(f, flags)) {
-        bool rflow = false;
-        alproto = AppLayerProtoDetectPPGetProto(f, buf, buflen, ipproto, flags, &rflow);
+        *reverse_flow = false; /* to be safe if it was changed by PM */
+        alproto = AppLayerProtoDetectPPGetProto(f, buf, buflen, ipproto, flags, reverse_flow);
         if (AppProtoIsValid(alproto)) {
-            if (rflow) {
-                *reverse_flow = true;
-            }
-            goto end;
+            SCReturnUInt(alproto);
         }
     }
 
     /* Look if flow can be found in expectation list */
     if (!FLOW_IS_PE_DONE(f, flags)) {
+        *reverse_flow = false; /* to be safe if it was changed by PP */
```

Comment:
Not sure about this safety check... not sure it is needed...
Maybe we should rather have a DEBUG_VALIDATION like `DEBUG_VALIDATE_BUG_ON(*reverse_flow); // should not be set to true, otherwise we returned a valid alproto already`

Reply:
i'll inspect if the reversed flag is never set by the previous kind of the proto detection in the case of an unknown/failed result

Reply:
I've made an inspection of the corresponding code. Direction may be reported reversed only in a case of valid protocol returned. So the "to be safe" lines are not necessary. Now i believe even debug validation is not necessary.

At the same time i found a case when incorrect direction may be technically reported by the PM.
It's not in the scope of the current work. And most likely the case is not possible whith the currently registered set of protocol detection patterns. I can not provide a reproducer. The following is a result of code review only.

The PM technically detects a set of protos but only first one is used. Let's see at app-layer-detect-proto.c from line 1407
        uint16_t pm_matches = AppLayerProtoDetectPMGetProto(
                tctx, f, buf, buflen, flags, pm_results, reverse_flow);
        if (pm_matches > 0) {
            DEBUG_VALIDATE_BUG_ON(pm_matches > 1);
            alproto = pm_results[0];

DEBUG_VALIDATE_BUG_ON exists but the whole part of code looks confusing. Why do we use an array of matches when only first is expected? And i believe without debug a combination of patterns is possible when more than one is returned and some of the following damaged the 'reverse_flow'.

It looks a bit confusing why PMGetProtoInspect tries to traverse all patterns (from line 294 of app-layer-detect-proto.c) while its client expects the one and only. And it has even DEBUG_VALIDATE_BUG_ON check.

Reply:
Thanks for the investigation

> Now i believe even debug validation is not necessary.

It may be useful in the future, if some new code tries to break this assertion

> Why do we use an array of matches when only first is expected?

Because we use some generic code for multi-pattern matching. The debug validation is here so that fuzzing can find if there is one way to find mutliple protocols with patterns... which would lead likely to bad things...

Reply:
I see.
I prepared a version with the debug validation - https://github.com/OISF/suricata/pull/11644

Regarding the array of matches - I still don't quite understand why we do not break from the loop started at app-layer-detect-proto.c:294 when the pattern is found.
Well, I did not inspect the behavior fully and it's not in the scope of current change. Moreover, I hope the condition with multiple patterns detected is impossible with the current set of registered protocol detection patterns.

Reply:
> Moreover, I hope the condition with multiple patterns detected is impossible with the current set of registered protocol detection patterns.

That is the important point indeed...